app/models/concerns/waitlistable.rb

```diff
@@ -63,7 +54,34 @@ def waitlistable?
     end
 
     def waiting_list_position
-      @waiting_list_position ||= self.waiting_list&.position(self)
+      self.tracked_waitlist_position || self.waiting_list&.position(self)
+    end
+
+    def waiting_list_position?
+      self.waiting_list_position.present?
+    end
+
+    def waitlist_position_changed?
+      self.tracked_waitlist_position.present?
+    end
+
+    def waiting_list_position=(target_position)
+      self.tracked_waitlist_position = target_position
+
+      self.apply_to_waiting_list(target_position) if self.persisted? && waiting_list_persisted?
+    end
+
+    private def apply_to_waiting_list(target_position)
+      should_add = self.waitlistable? && target_position.nil?
+      should_move = self.waitlistable? && target_position.present?
+      should_remove = !self.waitlistable? && target_position.present?
+
+      self.waiting_list.add(self) if should_add
+      self.waiting_list.move_to_position(self, target_position) if should_move
+      self.waiting_list.remove(self) if should_remove
+
+      # Now that we definitely updated the "real" WL, we can forget about our tracking safely.
+      self.clear_tracked_waitlist_position!
```

Comment:
Explore whether we can make this conditional on a certainty of whether _something_ happened to persist a change to the database in lines 79-81

Reply:
Done! https://github.com/thewca/worldcubeassociation.org/pull/11173/commits/6586e164f2e5d0f0b9c2b50d6932486d85c03581

Reply:
And improved! https://github.com/thewca/worldcubeassociation.org/pull/11173/commits/2158abdd6acbaed013e4b061fa5162def7ada4cb